### PR TITLE
Interop support for the Configurator API and all plugins

### PR DESCRIPTION
**Pull request by jeuxjeux20**
It's currently hard to get anything — such as a window, at least — popping on your screen with the C API. This PR implements the entire Configurator API and all of its plugins. Making your dreams of a game engine usable in C a reality.

### What's implemented

- **`NrtConfigurator`**: methods to add plugin providers, as well as default systems
- **All variants of `IPluginProvider`**: (Graphics, Input, ResourceManagement, Windowing)
- **`GraphicsProvider`**: barebones with only a `Destroy` method
- **`IWindowingDevice`**: with all methods (except `TearDown`), such as `ProcessMessages`, `GetShouldClose` and `WindowTitle` manipulation
- **`IInputDevice`**: with all methods (except `Initialise` and `Update`)
- **`ResourceLoader`**: only the `ResourceRootDirectory` property has been implemented

#### LifetimeExtender
The `AllObjects<T>` function uses a templated static variable for each type, which can maybe cause issues with multiple TUs. However, linkers should automatically choose only *one* static variable across multiple TUs: that behavior is called [vague linkage](https://gcc.gnu.org/onlinedocs/gcc/Vague-Linkage.html).

#### NrtConfigurator
The `AddXPluginProvider` methods use `Lifetime::Find` to get a `shared_ptr` of a plugin provider. This method uses linear search on all `shared_ptr`s created by the C API. It shouldn't be an issue considering we don't spawn thousands of PluginProviders, but it's something to note.

#### NrtInputActionList
I choose to implement a span of `NovelRT::Input::InputAction` instances with a struct that mimics an actual `span`:  
```cpp
typedef struct {
    NrtInputActionHandle begin;
    NrtInputActionHandle end;
    intptr_t increment; // sizeof(InputAction)
} NrtInputActionList;
```
Contrary to iterators, they demand more "discipline" to be used, but they are much more efficient for performance (no allocation, no need to call an externally-linked method every time), easier to implement (it's just one struct!) and it actually makes sense for a span.
The `intptr_t Nrt_InputActionList_IncrementSize()` method is there if you want to make it a constant. (it's also pretty cool for a JIT!)

### CODE REVIEW

File: src/NovelRT.Interop/Graphics/NrtGraphicsProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/Graphics/NrtGraphicsProvider.cpp:28]:
```suggestion
    if (!Lifetime::Release(reinterpret_cast<GraphicsProvider*>(provider)))
    {
        Nrt_setErrMsgIsAlreadyDeletedOrRemovedInternal();
        return NRT_FAILURE_ALREADY_DELETED_OR_REMOVED;
    }
    
    return NRT_SUCCESS;
```

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (RubyNova) [inline src/NovelRT.Interop/LifetimeExtender.h:56]: Since this is C++ please add the correct qualifiers. I can't rationalise if this throws or mutates or what. :D

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (RubyNova) [inline src/NovelRT.Interop/LifetimeExtender.h:39]: does this not work?

```suggestion
        return nullptr;
```

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (RubyNova) [inline src/NovelRT.Interop/LifetimeExtender.h:22]: I think you might need `std::forward` not `std::move` here.

File: src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp:59]: Out of curiosity, what does the `auto&&` achieve here as opposed to just `auto`?

File: src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp:109]: Out of curiosity, what does the auto&& achieve here as opposed to just auto?

File: src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp:84]: Out of curiosity, what does the auto&& achieve here as opposed to just auto?

File: src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp:34]: Out of curiosity, what does the auto&& achieve here as opposed to just auto?

File: src/NovelRT.Interop/PluginManagement/NrtIGraphicsPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIGraphicsPluginProvider.cpp:28]:
```suggestion
    if (!Lifetime::Release(reinterpret_cast<IGraphicsPluginProvider*>(plugin)))
    {
        Nrt_setErrMsgIsAlreadyDeletedOrRemovedInternal();
        return NRT_FAILURE_ALREADY_DELETED_OR_REMOVED;
    }
    
    return NRT_SUCCESS;
```

File: src/NovelRT.Interop/PluginManagement/NrtIGraphicsPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIGraphicsPluginProvider.cpp:46]: Out of curiosity, what does the auto&& achieve here as opposed to just auto?

File: src/NovelRT.Interop/PluginManagement/NrtIInputPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIInputPluginProvider.cpp:28]:
```suggestion
    if (!Lifetime::Release(reinterpret_cast<IInputPluginProvider*>(plugin)))
    {
        Nrt_setErrMsgIsAlreadyDeletedOrRemovedInternal();
        return NRT_FAILURE_ALREADY_DELETED_OR_REMOVED;
    }
    
    return NRT_SUCCESS;
```

File: src/NovelRT.Interop/PluginManagement/NrtIInputPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIInputPluginProvider.cpp:45]: Out of curiosity, what does the auto&& achieve here as opposed to just auto?

File: src/NovelRT.Interop/PluginManagement/NrtIResourceManagementPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIResourceManagementPluginProvider.cpp:46]: Out of curiosity, what does the auto&& achieve here as opposed to just auto?

File: src/NovelRT.Interop/PluginManagement/NrtIResourceManagementPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIResourceManagementPluginProvider.cpp:28]:
```suggestion
    if (!Lifetime::Release(reinterpret_cast<IResourceManagementPluginProvider*>(plugin)))
    {
        Nrt_setErrMsgIsAlreadyDeletedOrRemovedInternal();
        return NRT_FAILURE_ALREADY_DELETED_OR_REMOVED;
    }
    
    return NRT_SUCCESS;
```

File: src/NovelRT.Interop/PluginManagement/NrtIWindowingPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIWindowingPluginProvider.cpp:28]:
```suggestion
    if (!Lifetime::Release(reinterpret_cast<IWindowingPluginProvider*>(plugin)))
    {
        Nrt_setErrMsgIsAlreadyDeletedOrRemovedInternal();
        return NRT_FAILURE_ALREADY_DELETED_OR_REMOVED;
    }

    return NRT_SUCCESS;
```

File: src/NovelRT.Interop/PluginManagement/NrtIWindowingPluginProvider.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/PluginManagement/NrtIWindowingPluginProvider.cpp:45]: Out of curiosity, what does the auto&& achieve here as opposed to just auto?

File: src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp:35]: We are no longer using the _s variants. Please put in the explicit checks required to use the function that exists in all places for what you require.

File: src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp:75]:
```suggestion
    if (!Lifetime::Release(reinterpret_cast<ResourceLoader*>(loader)))
    {
        Nrt_setErrMsgIsAlreadyDeletedOrRemovedInternal();
        return NRT_FAILURE_ALREADY_DELETED_OR_REMOVED;
    }

    return NRT_SUCCESS;
```

File: src/NovelRT.Interop/Windowing/NrtIWindowingDevice.cpp
Comment (RubyNova) [inline src/NovelRT.Interop/Windowing/NrtIWindowingDevice.cpp:51]: We are no longer using the _s variants. Please put in the explicit checks required to use the function that exists in all places for what you require.

**Review by RubyNova — CHANGES_REQUESTED** [review]
Overall a good first pass, definitely is going to require some adjustments but nothing catastrophic, also a few questions from me on some of the bits and pieces.

File: include/NovelRT.Interop/Graphics/NrtGraphicsTypedefs.h
Comment (jeuxjeux20) [reply include/NovelRT.Interop/Graphics/NrtGraphicsTypedefs.h:7]: Just added an include guard and another type, to make it ressemble other `NrtXTypedefs.h` files.

File: include/NovelRT.Interop/Windowing/NrtIWindowingDevice.h
Comment (jeuxjeux20) [reply include/NovelRT.Interop/Windowing/NrtIWindowingDevice.h:17]: It's quite hard currently to have consistent `char*` string allocation conventions, since `IWindowingDevice::GetWindowTitle` returns an `std::string` which inevitably copies the string, but other APIs don't return a string by copy and instead return it by reference (such as `Nrt_getLastError`).

Comment (RubyNova) [reply include/NovelRT.Interop/Windowing/NrtIWindowingDevice.h:17]: Possible bug in C++ then?

File: src/NovelRT.Interop/Ecs/NrtConfigurator.cpp
Comment (jeuxjeux20) [reply src/NovelRT.Interop/Ecs/NrtConfigurator.cpp:58]: Well, I should probably add checks in there...

File: src/NovelRT.Interop/Ecs/NrtConfigurator.cpp
Comment (jeuxjeux20) [reply src/NovelRT.Interop/Ecs/NrtConfigurator.cpp:146]: This should "move" the created scheduler into the heap, using the default move constructor. I don't think there's any issue there (Also, works on my machine)

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (jeuxjeux20) [reply src/NovelRT.Interop/LifetimeExtender.h:22]: Isn't `std::forward` reserved for perfect forwarding? For cases where you have `T&&` in a template (and not `std::shared_ptr<T>&&`).

Here, `std::move` just conducts the moving operation, because `emplace_pack(object)` would still consider `object` as an l-value.

File: src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp
Comment (jeuxjeux20) [reply src/NovelRT.Interop/PluginManagement/NrtDefaultPluginSelector.cpp:59]: Nothing much, using `auto` would lead to the same behavior. I liked `auto&&` because it sort of says that "this value is temporary and will be moved".

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (RubyNova) [reply src/NovelRT.Interop/LifetimeExtender.h:22]: Can't you perfect forward the pointer here? its a `T&&`

File: src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp
Comment (jeuxjeux20) [reply src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp:35]: Wouldn't it be better to create a function in `NrtMisc.h` containing this piece of code (with strcpy) for copying strings?

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (RubyNova) [reply src/NovelRT.Interop/LifetimeExtender.h]: I don't have time to check that - can emplace throw? can erase throw? I'm mostly trusting you to get it right here

Comment (jeuxjeux20) [reply src/NovelRT.Interop/LifetimeExtender.h]: Nothing throws!

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (RubyNova) [inline src/NovelRT.Interop/LifetimeExtender.h:30]: Can't this be `const`?

**Review by RubyNova — CHANGES_REQUESTED** [review]

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (jeuxjeux20) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: `WithPluginProvider` requires a non-const `std::shared_ptr<T>` so I don't think so.

Comment (RubyNova) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: uh, incorrect. The function does not mutate `this`. This method can be `const`. I'm pretty sure.
```suggestion
    template<typename T> [[nodiscard]] std::shared_ptr<T> Find(T* object) const noexcept
```

Comment (jeuxjeux20) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: But it's a method contained outside of a class, so it's not possible?

From C++ reference:
> cv - const/volatile qualification, only allowed in non-static member function declarations

Comment (RubyNova) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: I see. I thought these members were in a class.

How come they aren't?

Comment (jeuxjeux20) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: Thought that making a class would be overkill, what's the point of making a fully static class when C++ has functions with namespaces?

Comment (RubyNova) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: It's probably my C# brain talking. I'm not sure if there's a benefit.

File: src/NovelRT.Interop/Input/NrtInputAction.cpp
Comment (capnkenny) [reply src/NovelRT.Interop/Input/NrtInputAction.cpp:21]: If this were to be a `constexpr` function in C++, I believe it would become compile-time and therefore work as a constant (since `sizeof` is considered compile-time as well), no? Without that I agree - it'd be a runtime function and wouldn't provide the same benefit as expected.

Comment (RubyNova) [reply src/NovelRT.Interop/Input/NrtInputAction.cpp:21]: This is C

constexpr does not exist

:^)

Comment (RubyNova) [reply src/NovelRT.Interop/Input/NrtInputAction.cpp:21]: Also P/Invoke will not care for your constexpr :D

Comment (capnkenny) [reply src/NovelRT.Interop/Input/NrtInputAction.cpp:21]: Well its getting exposed as C, but its a c++ function 😛 (otherwise this file should end in `.c`)
Anyways if C# isn't going to care for it (how would Java then???), then either way it does appear to be moot

File: include/NovelRT.Interop/Windowing/NrtIWindowingDevice.h
Comment (capnkenny) [reply include/NovelRT.Interop/Windowing/NrtIWindowingDevice.h:17]: Maybe? In places that we have a `std::string`, i would thing we should be returning using `.c_str()` which gives us a `const char*`...

It's also odd because we don't use it much, but the entirety of the error handling API calls upon pre-defined `const char*`s

File: src/NovelRT.Interop/Input/NrtInputAction.cpp
Comment (RubyNova) [reply src/NovelRT.Interop/Input/NrtInputAction.cpp:21]: You can't make a C compliant function constexpr even if it is internally C++

File: src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp
Comment (capnkenny) [reply src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp:35]: Not sure when we decided this, @RubyNova ? It's still used in two other places and I couldn't find history of it - please open an issue to cover the spots where this is still used if that's the case?

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (capnkenny) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: Considering that this appears to be used across a few classes, I think this makes sense, in which case `const` would fall over at compile-time since there's no class for it.

@RubyNova if you're cool with it being outside of a class, then I think this is fine.

**Comment by capnkenny** [comment]
Note: may need to squash commits upon approval

File: src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp
Comment (RubyNova) [reply src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp:35]: We already have an open issue, see #388

File: src/NovelRT.Interop/LifetimeExtender.h
Comment (RubyNova) [reply src/NovelRT.Interop/LifetimeExtender.h:30]: C API is not my code owner area primarily - this is why I pinged you in. Your say is the final one here I think.

File: include/NovelRT.Interop/Input/NrtNovelKey.h
Comment (capnkenny) [inline include/NovelRT.Interop/Input/NrtNovelKey.h:15]: @RubyNova just to confirm, should we be directly outputting a `const char*` here? Or should we be returning a `NrtResult` and using a `const char**` parameter? We seem to use output parameters in other places but its not consistent.

File: include/NovelRT.Interop/Windowing/NrtIWindowingDevice.h
Comment (capnkenny) [reply include/NovelRT.Interop/Windowing/NrtIWindowingDevice.h:17]: Also another place to point out - direct output here? Or should we shift to a `NrtResult` return with an output parameter?

File: src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp
Comment (capnkenny) [reply src/NovelRT.Interop/ResourceManagement/NrtResourceLoader.cpp:35]: Ah looks like someone's been assigned it already - cool beans

File: include/NovelRT.Interop/Ecs/Graphics/NrtDefaultRenderingSystem.h
Comment (RubyNova) [reply include/NovelRT.Interop/Ecs/Graphics/NrtDefaultRenderingSystem.h:19]: This should be part of @NullSapphire 's work, see #405

File: src/NovelRT.Interop/Input/NrtInputAction.cpp
Comment (jeuxjeux20) [reply src/NovelRT.Interop/Input/NrtInputAction.cpp:21]: > I think C# would need to define its own constant. This function is at runtime. I don't entirely follow how this would work.

That's the problem. C# *can't* guess the size of `InputAction`. This is why it can make it a sort-of constant:
```cs
public static readonly nuint Size = Nrt_InputActionList_IncrementSize();
```

File: include/NovelRT.Interop/Input/NrtNovelKey.h
Comment (RubyNova) [reply include/NovelRT.Interop/Input/NrtNovelKey.h:15]: We originally agreed for getter only noexcept methods (so, pure property getters) we'd just do this - but that was a long time ago. You, as code owner, might feel this is no longer a good idea; but I think this is how it works.

File: include/NovelRT.Interop/Ecs/Graphics/NrtDefaultRenderingSystem.h
Comment (capnkenny) [reply include/NovelRT.Interop/Ecs/Graphics/NrtDefaultRenderingSystem.h:19]: Just confirming - ty!

File: src/NovelRT.Interop/Windowing/NrtIWindowingDevice.cpp
Comment (capnkenny) [inline src/NovelRT.Interop/Windowing/NrtIWindowingDevice.cpp:43]: If this doesn't throw in C++ land, can we not use `.c_str()` to handle the allocation for us and return a `const char*` instead?

Comment (jeuxjeux20) [reply src/NovelRT.Interop/Windowing/NrtIWindowingDevice.cpp:43]: That won't work. `GetWindowTitle` returns a `std::string` by value which copies the string onto the stack (this is what happens in this line). So using `c_str()` here will return a pointer to a string that's going to deleted as soon as the function ends.

File: src/NovelRT.Interop/Input/NrtInputAction.cpp
Comment (RubyNova) [reply src/NovelRT.Interop/Input/NrtInputAction.cpp:21]: Gotcha. Thanks. :D

File: src/NovelRT.Interop/Windowing/NrtIWindowingDevice.cpp
Comment (RubyNova) [reply src/NovelRT.Interop/Windowing/NrtIWindowingDevice.cpp:43]: Ah, now I remember.

I think this is going to require more thought, might be out of scope for this PR otherwise I think this change will get held up too long.

**Comment by RubyNova** [comment]
Update for this - I've gone ahead and implemented the resource loader APIs. I don't know how it affects your plugin support, but please review it.

**Comment by jeuxjeux20** [comment]
Update for this - I did the changes and adapted the existing code to the new APIs, the only thing missing is... the tests.

**Comment by RubyNova** [comment]
If you can't test some of it that's fine, just test what you can, we need to get this moving.

**Review by RubyNova — APPROVED** [review]

**Pull request merged**
